Add Layout Editor Decorations to client-side web panels

geowar1 opened the pull request:
PR for #7360

geowar1 commented:
Work-in-progress: Finished with the decoration elements/attributes import; will work on actual drawing code next.

geowar1 commented:
FYI: Bugfix near line 1451: "widget.connectdname" ==> "widget.connectcname".

geowar1 commented:
Also finished $drawBridge code; ready to merge! (removing WIP!)

pabender commented:
@geowar1 codeclimate is pointing out a large number of code smell/complexity/maintainability issues in this code.  Those should be addressed.

Also, both codeclimate and coveralls show that you are only covering 2% of the new java code in this pull request.  We have a 50% threshold for new code coverage set on codeclimate.

geowar1 commented:
So there are 7 functions with more than 4 arguments. The limit in JavaScript is 255. Do you really expect me to marshal the extra parameters into & then out of an array just to reduce them all to 4 arguments? I’m pretty sure that I’m not going to do that.

geowar1 commented:
There are 6 functions with “Cognitive Complexity” greater than 5… I could break them into smaller functions… but then I’d have to pass even more parameters and write even more test cases. They do what they’re intended to do. Rewriting them to reduce their “Cognitive Complexity” would actually make the code more complicated… not less.

geowar1 commented:
All the “similar code” may be similar… but they’re not identical. Best I could do is write a function to replace two of the five lines in four places… and it would need more than four parameters (to be different). (And would require writing even more test cases.) That’s really not worth the extra work.

geowar1 commented:
I’ll look into writing new test cases tomorrow.

geowar1 commented:
Where do you write the test cases for web/js/panel.js?

pabender commented:
@geowar1 I don’t see your argument for how breaking the methods up into smaller pieces requires more arguments.  The pieces you break the values into shouldn’t need all the parameters passed to the parent.

The number of parameters flags are really questioning how the data structures are organized.  If you find yourself repeatedly passing some particular values together, it may be worthwhile to build a structure for that.

In my opinion the cognitive complexity issue is more important than the number of parameters.  
Cognitive complexity is really measuring the cyclic complexity of any given piece of code.  That counts every break in the linear flow of the method.  This really is encouraging you to break the code up into smaller pieces which are easier to understand on their own AND are easier to unit test ( because each piece does less ).  This means the code is more maintainable.  Another trick is that if you give these new methods ( which can be private ) meaningful names, you can actually make the calling function more understandable.

pabender commented:
> Where do you write the test cases for web/js/panel.js?

There are cucumber tests, but that probably isn’t the right spot.

@rhwood what infrastructure do we have in place for unit testing JavaScript?

geowar1 commented:
Using a data structure to avoid using parameters just means marshaling those parameters into and then back out of those data structures before calling and then inside the function. That just adds unnecessary complexity compared to just passing them as parameters. If the panel.js code was class based then it might make sense to use member variables (in which case all class methods could be parameter less). But panel.js appears to be strait C type functions.

rhwood commented:
We don’t have any specific infrastructure for unit testing JavaScript at this point. I would have suggesting using the Java ScriptEngine for that, but since the JavaScript ScriptEngine (Nashorn) is officially deprecated by Oracle, I don’t want to introduce a dependency on that.

However, for these, cucumber tests that test for very specific behaviors are best, since we want to inspect not just that the JavaScript is tested, but that it has the same result in both Chrome and Firefox, but the tests need to get to the level of detail in the HTML DOM that they are testing that (for example) a tunnel is drawn correctly, since this code needs to deal with differences between those browsers.

If we were to introduce JavaScript testing, I’d use Karma within maven to avoid introducing another toolchain (npmjs) that would need to be installed and maintained by every developer. 

> On Sep 24, 2019, at 23:23, Paul Bender <notifications@github.com> wrote:
> 
> Where do you write the test cases for web/js/panel.js?
> 
> There are cucumber tests, but that probably isn’t the right spot.
> 
> @rhwood <https://github.com/rhwood> what infrastructure do we have in place for unit testing JavaScript?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/JMRI/JMRI/pull/7437?email_source=notifications&email_token=AACISEDXVKNWZI4FL3JWOQLQLLKRLA5CNFSM4IY3JIUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7QOYOA#issuecomment-534834232>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AACISEF6ZBBBFHYJPY3OH63QLLKRLANCNFSM4IY3JIUA>.
>

rhwood commented:
On the other hand, passing an object (widget) and two objects derived from it (ep1 and ep2) which are stored in an array (gPts), mean you can do the array lookup whenever you need to.

In other words, since you can retrieve ep1 and ep2 from gPts using the attributes of widget that are the keys to retrieving them only when needed, you do not pass them around.

> On Sep 25, 2019, at 01:40, George Warner <notifications@github.com> wrote:
> 
> Using a data structure to avoid using parameters just means marshaling those parameters into and then back out of those data structures before calling and then inside the function. That just adds unnecessary complexity compared to just passing them as parameters. If the panel.js code was class based then it might make sense to use member variables (in which case all class methods could be parameter less). But panel.js appears to be strait C type functions.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/JMRI/JMRI/pull/7437?email_source=notifications&email_token=AACISEAQMOQEQBRA3LMX2UTQLL2UHA5CNFSM4IY3JIUKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7QVOCA#issuecomment-534861576>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AACISEFNSLZSBQN2XWGU6GTQLL2UHANCNFSM4IY3JIUA>.
>

geowar1 commented:
Good example (retrieving ep1 & ep2 from $widget). The original code pulls out ep1 & ep2 tests if they exist (aborting the drawing if not) and then makes sure that they are numbers instead of strings (so comparisons later are numerical not strings). Is duplicating all that code in ever routine that uses ep1 & ep2 really a smart thing to do? If that code ever has to be changed then all the duplicates will have to be searched down and changed also (increasing the chance of missing one of the duplicates and introducing new bugs unintentionally). (Note: A single function could be written to perform that task and all the duplicate locations could be changed to use that function instead.) And then there's the performance issue: there's an overhead of calling that duplicated code over and over again with the initially extracted and converted values could just be reused.

pabender commented:
@geowar1 that is an argument for using an object for the common processing.  That object provides methods for any common tasks.

On testing of the JavaScript with cucumber, if you have an example panel that includes all of the decorations, we can add a test to load it during the cucumber tests very quickly ( it takes adding two lines to one file ).

geowar1 commented:
Just a note about code complexity: Most (if not all) of the new code was copied directly from the Java sources and then converted to JavaScript. Rewriting the javascript to reduce code complexity would break the similarity of the Java and JavaScript code meaning the code would be more difficult to maintain if (when) a change has to be made to both. (Counter argument here: the Java version should probably be made less complex (in a parallel manor to the JavaScript) as well.) In the Java version a decorations class could be added with specific subclasses for arrows, end bumpers, bridges and tunnels. Common code could be in the super (base) class.

geowar1 commented:
I have a panel with all track segment <-> decoration combinations that I use for (manual visual) testing… so that's QED. Is there a way to compare the resulting image with a presaved image to verify that a change to the drawing code doesn't break the visual representation? (If (when) it does a new presaved image could be generated that reflects intentional changes.)

geowar1 commented:
Downside to comparing rendered to presaved image: It's an "all or nothing" comparison. IOW: It won't tell me what it found different, just that the images are different. I could break the one panel of all track segment <-> decorations combinations into individual panels (one for each comparison). That would give better detail as to what failed (but would be a lot of work).

geowar1 commented:
> @pabender I don’t see your argument for how breaking the methods up into smaller pieces requires more arguments.
There aren't any methods; the JavaScript panel drawing code is written as procedural C… no classes. Therefore I can't use object variables to store values in between calling methods. I can only pass values as parameter into functions. (Unless I rewrite both the Java and JavaScript to use a TrackDecorations class?) Plus given the complexity of what the code does breaking it up into (sub)functions would just make it harder to figure out what the code does; it would also require a lot more duplication of code (leading to future bugs).

pabender commented:
> > @pabender I don’t see your argument for how breaking the methods up into smaller pieces requires more arguments.
> 
> There aren't any methods; the JavaScript panel drawing code is written as procedural C… no classes. Therefore I can't use object variables to store values in between calling methods. I can only pass values as parameter into functions. (Unless I rewrite both the Java and JavaScript to use a TrackDecorations class?) Plus given the complexity of what the code does breaking it up into (sub)functions would just make it harder to figure out what the code does; it would also require a lot more duplication of code (leading to future bugs).

So you are arguing that taking something that is too complex to understand and breaking it up into more understandable pieces makes it less understandable.  That isn’t a logical conclusion.

Also, even procedural C uses functions and data structures.  Also, just because the JavaScript doesn’t use classes now doesn’t mean it shouldn’t.

pabender commented:
> Downside to comparing rendered to presaved image: It's an "all or nothing" comparison. IOW: It won't tell me what it found different, just that the images are different. I could break the one panel of all track segment <-> decorations combinations into individual panels (one for each comparison). That would give better detail as to what failed (but would be a lot of work).

The starting point is does the panel actually load without errors.  The panel you talk about should be loaded during the testing of the layout editor code too.

pabender commented:
> Just a note about code complexity: Most (if not all) of the new code was copied directly from the Java sources and then converted to JavaScript. Rewriting the javascript to reduce code complexity would break the similarity of the Java and JavaScript code meaning the code would be more difficult to maintain if (when) a change has to be made to both. (Counter argument here: the Java version should probably be made less complex (in a parallel manor to the JavaScript) as well.) A decorations class could be added with specific subclasses for arrows, end bumpers, bridges and tunnels. Common code could be in the super (base) class.

The java code needs to be refactored, but there is a problem with that.  There are insufficient tests to tell if a refactoring broke anything.  This is one reason why the code you copied from is effectively unmaintainable.

Also, this isn’t just me saying that.  Code climate also shows us that.  Just as two examples in that package, See:
https://codeclimate.com/github/JMRI/JMRI/java/src/jmri/jmrit/display/layoutEditor/LayoutEditor.java
And
https://codeclimate.com/github/JMRI/JMRI/java/src/jmri/jmrit/display/layoutEditor/LayoutTurnout.java

pabender commented:
To make your panel load during the cucumber tests, you need to add your panel to git, then edit:
https://github.com/JMRI/JMRI/blob/master/java/acceptancetest/features/web/WebPanel.feature

Add the file for both firefox and chrome please.

geowar1 commented:
@pabender I never said that it was too complex to understand… just too complex to break into smaller chunks (too many linear dependencies to maintain a known drawing state without massive overhead of constantly saving & restoring same). If I had it to do over I’d write the decorations as a class in the Java sources; that way I could easily parallel the JavaScript sources to match. But I didn’t. And at this late date I can’t justify refactoring one without the other. And that’s more work that I’m willing to do. I am willing to write as many test cases as will satisfy you (for both testing and code coverage stats).

pabender commented:
@geowar1 If code climate were only pointing out a couple of issues, then I could probably overlook that, but, code climate is pointing out 43 issues with the code.  That is too many.

Incidentally, the warnings that say function x has "too many lines" and "too high a cognitive complexity" both measure the fact that the method involved is difficult to understand.  If you fix one of those, you will end up fixing both of those.

I am not going to merge this PR until that is fixed (not that it matters at the moment, it isn't passing tests, so I can't merge the PR).

pabender commented:
both travis and appveyor a failing on a track segment test, which is something changed by this PR.

geowar1 commented:
So basically you’re freezing all future development on the Layout Editor because anyone touching it will have to uncomplexify it and the panel.js sources. No one (myself included) is going to be willing to do that.
JMRI pedantic peer review: 1, JMRI users: zero.

pabender commented:
> So basically you’re freezing all future development on the Layout Editor because anyone touching it will have to uncomplexify it and the panel.js sources. No one (myself included) is going to be willing to do that.

I am not actually saying that.  I am saying that if I am going to approve the PR, that NEW CODE is better at meeting standards than the old code.  You may have copied the java code to get the JavaScript working, but it is still new code to the codebase.

There is a reason we put the code climate checks in place, and it is to help guide us in making things better.  New features are good, but they also have to be maintainable.

That said, I am not the only one who is able to approve pull requests. Someone else may be willing to let this go in as us, but I am not.  Sorry.

pabender commented:
One more comment on this, I don’t think anyone should approve a pull request that adds more than 1400 untested lines of code to any project.  That is too much code to review by hand, so we must use the tools at hand to help us.  You added 1425 lines to panel.js.

Just taking your word that it is ready defeats the purpose of peer review, but that is basically what you are asking me to do.

Pugwash1 commented:
@geowar1
@pabender has a point about adding a large number of lines without any objective tests. 
Codeclimate can give some interesting insights into code, but is so subjective as to be useless out side of the academic world. It even breaks it's own rule  "too high a cognitive complexity", that my English teacher would have red circled and commented "Stephen, why do always hide what your saying from 99.9% of the population, stick with the Anglo Saxon!"

pabender commented:
> Codeclimate can give some interesting insights into code, but is so subjective as to be useless out side of the academic world. It even breaks it's own rule "too high a cognitive complexity", that my English teacher would have red circled and commented "Stephen, why do always hide what your saying from 99.9% of the population, stick with the Anglo Saxon!"

If I were writing the tool, I would have used cyclomatic complexity instead:
https://en.m.wikipedia.org/wiki/Cyclomatic_complexity

This is at least a portion of what is being measured.

geowar1 commented:
I haven't voiced any complaint about writing test cases. In fact I've explicated said that I'm willing to do that provided someone can tell me where to do so. @rhwood commented with "We don’t have any specific infrastructure for unit testing JavaScript at this point.". You added "if you have an example panel that includes all of the decorations, we can add a test to load it during the cucumber tests very quickly". As I stated, I'm perfectly willing to do that. Are you willing to accept the new panel.js code as tested if that file is included?
No, all my complaints are about code climates unrealistic conditions: No nested if statements? (too complex) no more than 4 parameters per function (/method)? No more than 26 lines per function? No more than 255 lines per source file? Ridiculous. That's what I'm objecting to. I challenge anyone to rewrite/refactor LayoutManager.java with those restrictions. It would have to be split into more than 42 source files. No sane developer is going to do that.

pabender commented:
travis headed restarted after an instance of #7465

pabender commented:
> No, all my complaints are about code climates unrealistic conditions: No nested if statements? (too complex) no more than 4 parameters per function (/method)? No more than 26 lines per function? No more than 255 lines per source file? Ridiculous. That's what I'm objecting to.

I don't see everything you mention as being unrealistic.  When I started programming, the expectation was that every method would fit on single 80 column wide 24 line terminal window.  That has always been my expectation, though sometimes even I don't hit that.

Some of the items code climate checks we are going to adjust, but not all of them.  (See #7463 and #7464).

> I challenge anyone to rewrite/refactor LayoutManager.java with those restrictions. It would have to be split into more than 42 source files. No sane developer is going to do that.

I am assuming you are talking about LayoutEditor, since I don't know what LayoutManager is.  

What codeclimate tells us about LayoutEditor.java, which is something that @bobjacobsen and I have both stated for a long time:  In it's present form it is unmaintainable.  i.e. the risk involved with fixing bugs in the code is large, because the class is large and complicated and there are insufficient tests of the current behavior.

But we have a bit of a chicken and egg problem.  We need more tests so we are more likely to catch new errors introduced into the code, but we also need to refactor the code so that it is more testable and more maintainable.   

The real issue is that it isn't JUST LayoutEditor.java in that package which is a problem.  That single package has more than 12% of the untested code in the entire project! (more than 24,000 untested lines  (see http://jmri.tagadab.com/jenkins/job/Development/job/JaCoCo/2381/jacoco/jmri.jmrit.display.layoutEditor/ ) out of a little under 200,000 total uncovered lines (see http://jmri.tagadab.com/jenkins/job/Development/job/JaCoCo/2381/jacoco/ ). 

Now where do the CodeClimate metrics come into play?  They give us some pointers into how we can reduce the number of lines of code in the package, by eliminating duplicate code, and where we should focus our energy to make the code more testable.  

Ultimately, that means LayoutEditor needs to be internally redesigned so that it is more efficient (it needs to be DRYed out).  The functionality needs to be broken out into smaller, more testable, pieces.  How many files is that going to take?  I don't know.  I just know the present situation isn't good, and I don't want that to spread to other parts of the code base.

Also, Thank you for adding the cucumber test that is a start to testing this code.

geowar1 commented:
Git now says I've touch 47 files… I've only touched 8. Gonna have to try again… Gerrrrrrrr…

geowar1 commented:
Closed and reopened as #7482.
